jaeger-core/src/test/java/com/uber/jaeger/ConfigurationTest.java

```diff
@@ -320,6 +339,490 @@ public void testPropagationValidFormat() {
     assertNotNull(textMap.get("uber-trace-id"));
   }
 
+  @Test
+  public void testEnsureCachedTracerInstance() {
+    Configuration configuration = new Configuration("the-service");
+    io.opentracing.Tracer tracer = configuration.getTracer();
+    assertEquals("Same instance of the tracer was expected", tracer, configuration.getTracer());
+  }
+
+  @Test
+  public void testEnsureNewTracerInstance() {
+    Configuration configuration = new Configuration("the-service");
+    io.opentracing.Tracer tracer = configuration.getTracerBuilder().build();
+    assertNotEquals("Different instance of the tracer was expected", tracer, configuration.getTracerBuilder().build());
+  }
+
+  @Test
+  public void testCloseWhenNoTracerWasBuilt() {
+    // this is a noop, but we ensure there's no NPE going on
+    new Configuration("the-service").closeTracer();
+  }
+
+  @Test
+  public void testClose() throws SenderException {
+    Sender sender = mock(Sender.class);
+    ReporterConfiguration reporterConfiguration = new ReporterConfiguration(sender);
+
+    Configuration configuration = Configuration
+        .builder("a-name")
+        .withReporterConfiguration(reporterConfiguration)
+        .build();
+
+    configuration.getTracer();
+    configuration.closeTracer();
+
+    verify(sender, times(1)).close();
+  }
+
+  @Test
+  public void testServiceName() {
+    Configuration configuration = Configuration.builder("the-service").build();
+    io.opentracing.Tracer tracer = configuration.getTracer();
+    assertEquals("the-service", configuration.getServiceName());
+    assertEquals("the-service", ((Tracer) tracer).getServiceName());
+  }
+
+  @Test
+  public void testServiceNameOnConstructor() {
+    Configuration configuration = new Configuration("the-service");
+    assertEquals("Expected a different service name", "the-service", configuration.getServiceName());
+
+    Tracer tracer = (Tracer) configuration.getTracer();
+    assertEquals("Expected a different service name", "the-service", tracer.getServiceName());
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void testNoServiceNameFails() {
+    Configuration.builder(null).build().getTracer();
+  }
+
+  @Test
+  public void testOverrideServiceName() {
+    Configuration configuration = Configuration
+        .builder("the-service")
+        .withServiceName("another-service")
+        .build();
+    assertEquals("Expected a different service name", "another-service", configuration.getServiceName());
+
+    Tracer tracer = (Tracer) configuration.getTracer();
+    assertEquals("Should have overridden the service name", "another-service", tracer.getServiceName());
+  }
+
+  @Test
+  public void testBuilderFromEnv() {
+    System.setProperty(Configuration.JAEGER_SERVICE_NAME, "the-service");
+    Tracer tracer = (Tracer) Configuration
+        .builder()
+        .fromEnv()
```

Comment:
Just to clarify the behaviour - so iany preceding configuration that may have been defined will be overwritten by the values obtained from env - and similarly it would be possible to override parts of the env config by using `with...` methods after the `fromEnv` but before the `build`?

Reply:
The idea is that people would be able to get everything from the env and tweak one or two values. It happens to work the other way around as well, but that's a nice side effect.